libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1027 @@
+      return NULL;
+    }
+
+  serial = json_node_get_int (serial_node);
+
+  detail_object = json_object_get_object_member (object, "detail");
+
+  if (detail_object == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (node);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected a 'detail' member in the autoinstall spec (at %s)", node_str);
+      return NULL;
+    }
+
+  flatpak_remote_ref = flatpak_remote_ref_from_action_detail (action_type, detail_object, &local_error);
+
+  if (!flatpak_remote_ref)
+    {
+      if (g_error_matches (local_error,
+                           EOS_UPDATER_ERROR,
+                           EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC))
+        {
+          g_autofree gchar *node_str = json_node_to_string (node);
+          g_propagate_prefixed_error (error,
+                                      local_error,
+                                      "Error parsing the 'detail' member for action (at %s) '%s': ",
+                                      action_type_str,
+                                      node_str);
+          return NULL;
+        }
+
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return NULL;
+    }
+  
+  return flatpak_remote_ref_action_new (action_type, flatpak_remote_ref, serial);
+}
+
+static gint
+sort_flatpak_remote_ref_actions (gconstpointer a, gconstpointer b)
+{
+  FlatpakRemoteRefAction **action_a = (FlatpakRemoteRefAction **) a;
+  FlatpakRemoteRefAction **action_b = (FlatpakRemoteRefAction **) b;
+
+  return (gint) (*action_a)->serial - (gint) (*action_b)->serial;
+}
+
+static JsonNode *
+parse_json_from_file (GFile         *file,
+                      GCancellable  *cancellable,
+                      GError       **error)
+{
+  g_autoptr(GFileInputStream) input_stream = NULL;
+  g_autoptr(JsonParser) parser = json_parser_new_immutable ();
+
+  input_stream = g_file_read (file, cancellable, error);
+
+  if (!input_stream)
+    return NULL;
+
+  if (!json_parser_load_from_stream (parser,
+                                     G_INPUT_STREAM (input_stream),
+                                     cancellable,
+                                     error))
+    return NULL;
+
+  return json_node_ref (json_parser_get_root (parser));
+}
+
+static const gchar *
+eos_updater_get_system_architecture_string (void)
+{
+  return eos_updater_get_envvar_or ("EOS_UPDATER_TEST_OVERRIDE_ARCHITECTURE",
+                                    flatpak_get_default_arch ());
+}
+
+static gboolean
+architecture_skip_filter_applies (JsonNode  *filter_value,
+                                  gboolean  *is_filtered,
+                                  GError   **error)
+{
+  JsonArray *skip_architectures_array = json_node_get_array (filter_value);
+  g_autoptr(GList) architecture_nodes = NULL;
+  g_autofree gchar *node_str = json_node_to_string (filter_value);
```

Comment:
Please move this inside the error blocks, so that we don't need to do that in the regular case.